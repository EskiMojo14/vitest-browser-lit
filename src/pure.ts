import type { LocatorSelectors } from 'vitest/browser'
import { utils } from 'vitest/browser'
import type { RenderOptions } from 'lit'
import { render as litRender } from 'lit'

const { debug, getElementLocatorSelectors } = utils

export interface ComponentRenderOptions extends RenderOptions {
  baseElement?: HTMLElement
  container?: HTMLElement
}

export interface RenderResult extends LocatorSelectors {
  container: HTMLElement
  baseElement: HTMLElement
  debug: typeof utils.debug
  unmount: () => void
  rerender: (template: unknown) => void
  asFragment: () => DocumentFragment
}

const containers = new Set<HTMLElement>()

export function render(
  template: unknown,
  {
    baseElement = document.body,
    container = baseElement.appendChild(document.createElement('div')),
    ...options
  }: ComponentRenderOptions = {},
): RenderResult {
  containers.add(container)
  litRender(template, container, options)
  return {
    container,
    baseElement,
    debug,
    unmount() {
      containers.delete(container)
      container.remove()
    },
    rerender(newTemplate: unknown) {
      litRender(newTemplate, container, options)
    },
    asFragment() {
      return document
        .createRange()
        .createContextualFragment(container.innerHTML)
    },
    ...getElementLocatorSelectors(container),
  }
}

export function cleanup(): void {
  for (const container of containers) {
    container.remove()
  }
  containers.clear()
}
