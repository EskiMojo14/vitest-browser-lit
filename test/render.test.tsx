import { html } from 'lit'
import { expect, test } from 'vitest'
import { page, userEvent } from '@vitest/browser/context'
import { render } from '../src/index'
import './fixtures/hello-world'
import './fixtures/counter'

test('renders simple component', async () => {
  const screen = render(html`<hello-world></hello-world>`)
  await expect.element(page.getByText('Hello World')).toBeVisible()
  expect(screen.container).toMatchSnapshot()
})
test('renders counter', async () => {
  const screen = render(html`<counter-element count="1"></counter-element>`)

  await expect.element(screen.getByText('Count is 1')).toBeVisible()
  await userEvent.click(screen.getByRole('button', { name: 'Increment' }))
  await expect.element(screen.getByText('Count is 2')).toBeVisible()
})
