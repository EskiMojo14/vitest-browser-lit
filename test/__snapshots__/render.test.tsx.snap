// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`renders simple component 1`] = `
<div>
  <!---->
  <hello-world>
    #shadow-root
      <div>
        Hello World
      </div>
  </hello-world>
</div>
`;
